add `@memmove` builtin

dweiller opened the pull request:
This PR adds a new builtin `@memmove(dest: T, src: T) void`. It is similar to `@memcpy` except it can be used for overlapping memory regions. The intended semantics are:

  - `src` and `dest` can each be a slice, pointer to array, or many-item pointer
  - at least one of `src` and `dest` must provide a length, and if they both do the lengths must be the same
  - ~unlike (the intended runtime, and current comptime) behaviour of`@memcpy`,~ `@memmove` requires the element type of `src` and `dest` to be coercible to each other in memory
  - the result of `@memmove(dest, src)` is that `dest[0..len]` contains bytes as if a element-by-element loop copied `src[0..len]` to an auxiliary buffer, and then a second loop copied from that buffer to `dest[0..len]`.

This builtin is currently using up the last ZIR tag.

Related: #767

nektro commented:
- should it be `@memmove(dest, src)` to match order of `@memcpy` ?
- does `@memmove` perform the same length equality assertion as `@memcpy` ?

dweiller commented:
> * should it be `@memmove(dest, src)` to match order of `@memcpy` ?
> 
>     * does `@memmove` perform the same length equality assertion as `@memcpy` ?

Whoops the order iis the same, I just wrote it wrong. It has the same length checks too - will update the OP.

alexrp commented:
> unlike (the intended runtime, and current comptime) behaviour of`@memcpy`, `@memmove` requires the element type of `src` and `dest` to be coercible to each other in memory

Just noting that we concluded in the discussion that `@memcpy` will be changed to only allow in-memory coercions, i.e. `coerceInMemoryAllowed`. So the builtins will be consistent on this point.

alexrp commented:
Additionally, per the `@memcpy` discussion, the addition of `@memmove` should include deprecation of `std.mem.copy(Backwards,Forwards)` (and later removal in 0.15.0).

dweiller commented on doc/langref.html.in:
```diff
@@ -5081,6 +5081,23 @@ fn cmpxchgWeakButNotAtomic(comptime T: type, ptr: *T, expected_value: T, new_val
       {#syntax#}std.crypto.secureZero{#endsyntax#}</p>
       {#header_close#}
 
+      {#header_open|@memmove#}
+      <pre>{#syntax#}@memmove(dest, source) void{#endsyntax#}</pre>
+      <p>This function copies bytes from one region of memory to another, but unlike
+      {#link|@memcpy#} the regions may overlap.</p>
+      <p>{#syntax#}dest{#endsyntax#} must be a mutable slice, a mutable pointer to an array, or
+        a mutable many-item {#link|pointer|Pointers#}. It may have any
+        alignment, and it may have any element type.</p>
+      <p>{#syntax#}source{#endsyntax#} must be a slice, a pointer to
+        an array, or a many-item {#link|pointer|Pointers#}. It may
+        have any alignment, and it may have any element type.</p>
+      <p>The {#syntax#}source{#endsyntax#} and {#syntax#}dest{#endsyntax#} must have the same element
+        type.</p>
+      <p>Similar to {#link|for#} loops, at least one of {#syntax#}source{#endsyntax#} and
+        {#syntax#}dest{#endsyntax#} must provide a length, and if two lengths are provided,
+        they must be equal.</p>
+      {#header_close#}
+
```
The docs should describe the semantics of the copy when the pointers alias. Rather than saying the elements must have the same type it should probably say something about coercible in memory.

dweiller commented:
I don't know what the aarch64-macos-debug CI issue is - it failed before and after a rebase for `@disableIntrinsics()`, but I can't really believe this PR is causing the problem.

dweiller commented:
Looks like the aarch64-macos-debug CI issue is unrelated, there are other PRs that are having the same failure.

andrewrk commented:
macho linker bug filed at #23010

andrewrk commented:
```
andy@bark ~/d/z/build-release (dweiller-memmove)> git rebase origin/master
Successfully rebased and updated refs/heads/dweiller-memmove.
andy@bark ~/d/z/build-release (dweiller-memmove)> /etc/profiles/per-user/andy/bin/fzf-tmux: line 22: tmux: commaandy@bark ~/d/z/build-release (dweiller-memmove)> git push -f git@github.com:dweiller/zig.git dweiller-memmove:memmove
Enumerating objects: 149, done.
Counting objects: 100% (149/149), done.
Delta compression using up to 32 threads
Compressing objects: 100% (82/82), done.
Writing objects: 100% (85/85), 11.15 KiB | 11.15 MiB/s, done.
Total 85 (delta 70), reused 0 (delta 0), pack-reused 0 (from 0)
remote: Resolving deltas: 100% (70/70), completed with 56 local objects.
To github.com:dweiller/zig.git
 ! [remote rejected]       dweiller-memmove -> memmove (permission denied)
error: failed to push some refs to 'github.com:dweiller/zig.git'
```

dweiller commented:
> ```
> andy@bark ~/d/z/build-release (dweiller-memmove)> git rebase origin/master
> Successfully rebased and updated refs/heads/dweiller-memmove.
> andy@bark ~/d/z/build-release (dweiller-memmove)> /etc/profiles/per-user/andy/bin/fzf-tmux: line 22: tmux: commaandy@bark ~/d/z/build-release (dweiller-memmove)> git push -f git@github.com:dweiller/zig.git dweiller-memmove:memmove
> Enumerating objects: 149, done.
> Counting objects: 100% (149/149), done.
> Delta compression using up to 32 threads
> Compressing objects: 100% (82/82), done.
> Writing objects: 100% (85/85), 11.15 KiB | 11.15 MiB/s, done.
> Total 85 (delta 70), reused 0 (delta 0), pack-reused 0 (from 0)
> remote: Resolving deltas: 100% (70/70), completed with 56 local objects.
> To github.com:dweiller/zig.git
>  ! [remote rejected]       dweiller-memmove -> memmove (permission denied)
> error: failed to push some refs to 'github.com:dweiller/zig.git'
> ```

I've rebased onto master (c06fecd466) and enabled the 'allow edits by maintainers' option so you shouldn't hit that error again.

andrewrk commented:
Thanks!

dweiller commented:
Another option could be to have a single `@copy` builtin put a third parameter of type `enum { no_alias, may_alias }`, since builtin's can be variadic, we could have `@copy(s1, s2)` default to one of them (`may_alias` might be best).
